### PR TITLE
Implement io.Seeker on unixio.DAGReader

### PR DESCRIPTION
**Pull request by whyrusleeping**
This PR makes jbenet's dream of seeking through movies hosted in ipfs a reality, kinda, i didnt implement the seeking part in the filesystem or the web interface, just on the object.

### CODE REVIEW

File: merkledag/merkledag.go [hunk 1]
```diff
@@ -169,41 +168,78 @@ func FindLinks(n *Node, k u.Key, start int) []int {
...
-				// NB: can occur in normal situations, with improperly formatted
-				// input data
-				log.Error("Got back bad block!")
-				break
-			}
-			is := FindLinks(root, blk.Key(), next)
-			for _, i := range is {
-				nodes[i] = nd
-			}
+func (ds *dagService) GetNodes(ctx context.Context, keys []u.Key) []NodeGetter {
```
Comment (jbenet) [inline merkledag/merkledag.go]: godoc on `GetNodes` for users

File: unixfs/io/dagmodifier_test.go [hunk 2]
```diff
@@ -38,7 +39,7 @@ func getNode(t *testing.T, dserv mdag.DAGService, size int64) ([]byte, *mdag.Nod
 		t.Fatal(err)
 	}
 
-	dr, err := NewDagReader(node, dserv)
+	dr, err := NewDagReader(context.TODO(), node, dserv)
```
Comment (jbenet) [inline unixfs/io/dagmodifier_test.go]: use `context.TODO()` when it's mean to (+ will be) be wired up to something else, and `context.Background()` when it's not.

File: unixfs/io/dagreader.go [hunk 3]
```diff
@@ -19,14 +20,29 @@ var ErrIsDir = errors.New("this dag node is a directory")
...
+	promises     []mdag.NodeGetter
 	linkPosition int
+	offset       int64
+
+	// Our context
+	ctx context.Context
+
+	// Context for children
+	fctx   context.Context
+	cancel func()
```
Comment (jbenet) [inline unixfs/io/dagreader.go]: - `fctx` isn't set anywhere.
- `cancel` corresponds to `ctx`
- when you want to embed the ctx as state in an object, with cancellation as the main purpose, try `goprocess`: 

```go
type DagReader struct {
  ...
  proc goprocess.Process
}

func (dr *DagReader) Close() error {
  return dr.proc.Close()
}
```

File: unixfs/io/dagreader.go [hunk 4]
```diff
@@ -19,14 +20,29 @@ var ErrIsDir = errors.New("this dag node is a directory")
 type DagReader struct {
 	serv         mdag.DAGService
 	node         *mdag.Node
-	buf          io.Reader
-	fetchChan    <-chan *mdag.Node
+	pbdata       *ftpb.Data
+	buf          ReadSeekCloser
+	promises     []mdag.NodeGetter
 	linkPosition int
+	offset       int64
```
Comment (jbenet) [inline unixfs/io/dagreader.go]: document. i would not feel comfortable jumping in and modifying. a quick comment next to each field, and a more thorough comment for the whole object describing how it operates, would be very helpful.

File: unixfs/io/dagreader.go [hunk 5]
```diff
@@ -133,28 +140,73 @@ func (dr *DagReader) Read(b []byte) (int, error) {
 	}
 }
 
-/*
+func (dr *DagReader) Close() error {
+	dr.cancel()
+	return nil
+}
+
 func (dr *DagReader) Seek(offset int64, whence int) (int64, error) {
```
Comment (jbenet) [inline unixfs/io/dagreader.go]: godoc on `Seek` describing how it seeks.

File: unixfs/io/dagreader.go [hunk 6]
```diff
@@ -133,28 +140,73 @@ func (dr *DagReader) Read(b []byte) (int, error) {
...
 	switch whence {
 	case os.SEEK_SET:
-		for i := 0; i < len(dr.node.Links); i++ {
-			nsize := dr.node.Links[i].Size - 8
-			if offset > nsize {
-				offset -= nsize
-			} else {
+		if offset < 0 {
```
Comment (jbenet) [inline unixfs/io/dagreader.go]: comments along the `os.SEEK_SET` part, so it's easier to follow:

```
    // seek past the data stored in this node.
    pb := dr.pbdata
    left := offset
    if int64(len(pb.Data)) > offset {
      dr.buf.Close()
      dr.buf = NewRSNCFromBytes(pb.GetData()[offset:])
      dr.linkPosition = 0
      dr.offset = offset
      return offset, nil
    } else {
      left -= int64(len(pb.Data))
    }
```
etc

File: importer/importer_test.go [hunk 7]
```diff
@@ -179,3 +181,171 @@ func TestIndirectBlocks(t *testing.T) {
...
+	}
+
+	err = arrComp(out, should[1:])
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSeekingConsistency(t *testing.T) {
+	nbytes := int64(128 * 1024)
```
Comment (jbenet) [inline importer/importer_test.go]: I'd love to see a stress test here. `128 * 1024 = 131072` which is smaller than our current blocksize. How about a:
```go
func TestSeekingStress(t *testing.T) {
  buf := make([]byte,  1MB) // we could reuse large random buffers across tests. init a read-only one in `init()`
  // read math/randomness into buf 

  buf2 := make([]byte, 1MB)
  for i := 0; i < 100; i++ {
    offset := rand.Intn(len(buf))
    length := offset + rand.Intn(len(buf) - offset)
    rs.Seek(offset, os.SEEK_SET)
    rs.Read(buf2[offset:length])
    if !bytes.Equal(buf2[offset:length], buf[offset:length]) {
      t.Error(...)
    }
  }
  

}
```

File: merkledag/merkledag.go [hunk 8]
```diff
@@ -26,7 +25,8 @@ type DAGService interface {
 
 	// GetDAG returns, in order, all the single leve child
 	// nodes of the passed in node.
-	GetDAG(context.Context, *Node) <-chan *Node
+	GetDAG(context.Context, *Node) []NodeGetter
+	GetNodes(context.Context, []u.Key) []NodeGetter
```
Comment (jbenet) [inline merkledag/merkledag.go]: explain the semantics of `NodeGetter` for the user.

**Comment by jbenet** [comment]
@whyrusleeping lots of comments up there o/

File: importer/importer_test.go (hunk 7 above)
Comment (whyrusleeping) [inline importer/importer_test.go]: Note that i set the blocksize to 500 in this test.

File: unixfs/io/dagreader.go (hunk 3 above)
Comment (whyrusleeping) [inline unixfs/io/dagreader.go]: Yeah, i would use goprocess, but the contexts main purpose is actually to derive children from for the sub-readers

File: importer/importer_test.go (hunk 7 above)
Comment (jbenet) [inline importer/importer_test.go]: ahhh, nice :+1: . would still like a large _random_ seek+read test. makes weird bugs harder to creep in

**Comment by whyrusleeping** [comment]
I dont know why the network test failed...

**Comment by jbenet** [comment]
LGTM!

**Comment by jbenet** [comment]
> I dont know why the network test failed...

it succeeded (see the `success`) but the script failed somehow. i've seen this before, not sure why.

File: importer/importer_test.go [hunk 9]
```diff
@@ -51,7 +54,7 @@ func testFileConsistency(t *testing.T, bs chunk.BlockSplitter, nbytes int) {
 		t.Fatal(err)
 	}
 
-	r, err := uio.NewDagReader(nd, dnp.ds)
+	r, err := uio.NewDagReader(context.TODO(), nd, dnp.ds)
```
Comment (jbenet) [inline importer/importer_test.go]: beep

File: importer/importer_test.go [hunk 10]
```diff
@@ -77,7 +80,7 @@ func TestBuilderConsistency(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	r, err := uio.NewDagReader(nd, dagserv)
+	r, err := uio.NewDagReader(context.TODO(), nd, dagserv)
```
Comment (jbenet) [inline importer/importer_test.go]: boop

File: importer/importer_test.go [hunk 11]
```diff
@@ -179,3 +182,216 @@ func TestIndirectBlocks(t *testing.T) {
...
+	u.NewTimeSeededRand().Read(should)
+
+	read := bytes.NewReader(should)
+	dnp := getDagservAndPinner(t)
+	nd, err := BuildDagFromReader(read, dnp.ds, dnp.mp, &chunk.SizeSplitter{500})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rs, err := uio.NewDagReader(context.TODO(), nd, dnp.ds)
```
Comment (jbenet) [inline importer/importer_test.go]: here

File: merkledag/merkledag.go [hunk 12]
```diff
@@ -169,41 +168,80 @@ func FindLinks(n *Node, k u.Key, start int) []int {
...
+	}, ch
+}
+
+type nodePromise struct {
+	cache *Node
+	recv  <-chan *Node
+	ctx   context.Context
+}
+
+type NodeGetter interface {
```
Comment (jbenet) [inline merkledag/merkledag.go]: Explain the semantics of NodeGetter for the user, because it's exposed to them

Comment (jbenet) [inline merkledag/merkledag.go]: (nifty struct btw)

**Comment by jbenet** [comment]
Awesome comments! thank you! :+1:

**Comment by jbenet** [comment]
LGTM RFM

**Pull request merged**